### ISSUE TITLE
Implement a virtual file system

### ISSUE THREAD

**Issue by HiPhish** [TurningWheel/Barony#79]
Currently Barony reads and writes all files from and into one directory. This means that log and config files are right next to game assets and modding means replacing official files. A virtual file system would allow defining a hierarchy of search paths which are then merged into one virtual file system that sorts out the *actual* file based on path priority.

As an example, assume the following setup: the player has the official build of the game which includes all the assets in `/usr/local/`. He downloads a mod that replaces some of the textures in `~/.local/share/Barony/my-mod`. The VFS would "merge" the two directories such that textures are first searched in the mod directory and if they are not found there the VFS would search for them in the official game directory. That way a mod does not have to overwrite anything and restoring the game is as easy as deleting the mod.

Written files like logs or configurations should be written to a separate directory as well, otherwise they "pollute" the game directory. That way the game can be installed for all users globally without having their savegames and settings step on each other. The game itself does not need to be concerned with this, the VFS will sort out the paths.

Another advantage is that it makes trying out different builds of the game easier. At the moment if I want to try a new build I have to move all the assets from one installation to another. With a VFS I could move all assets into one directory and then every build would know to find them there.

There is a VFS library called PhysicsFS: https://icculus.org/physfs/. Aside from working with directories, it can also abstract over archives like ZIP files. A mod could be just a ZIP file without any need to extract the archive. It has already been used with a couple of games like Dungeons of Dredmor, so it is known to work reliably.

What do you guys think, is it a good idea? If you want I could give it a spin when I find some free time, it looks like a task that could be managed by someone unfamiliar with the engine, since it's mostly wrapping the raw file operations with PhysicsFS counterparts.

**Comment by HiPhish** [comment]
Just as I finished typing up the above I updated from 2.0.4 to 2.0.5 on macOS and all my settings and savegames are gone. This would not have happened if the save and config files had not been stored inside the app bundle. I don't mind since I was only on level 2, but this sort of thing should not be happening.

**Comment by WALLOFJUSTICE** [comment]
Interesting, I can see this working if the case of map packs, as currently you backup the existing /map folder, and copy paste the provided /maps directory in to play. 

Can't conceptualise the savegames/configs use though... mind elaborating?

**Comment by HiPhish** [comment]
>Can't conceptualise the savegames/configs use though... mind elaborating?

Sure. Aside from the various read-directories there is also a write-directory. On Unix it could for example be `~/.config/Barony/`, then in your code you would have
```c
/* Set the write directory when the program starts */
PHYSFS_setWriteDir("~/.config/Barony/"); 

/* Write save file at some later point in time */
PHYSFS_uint32 save_size; 
void *save_data;

/* Note the relative file name */
PHYSFS_File *save_file = PHYSFS_openWrite("savegame.dat");
PHYSFS_write(save_file, save_data, 1, save_size);
```

All files written like this will be written to the `~/.config/Barony` directory, we only need to set up the directory once when the game starts and then we can keep all file names and file path relative just as they are now.

**Comment by WALLOFJUSTICE** [comment]
Right. For implementation, with new Barony executables I would like individual save files written depending on which .zip file the executable was run. Does the virtual save file take priority over the default in the case where there are 2 identical file names?

**Comment by HiPhish** [comment]
>depending on which .zip file the executable was run
I don't understand what that means.

**Comment by WALLOFJUSTICE** [comment]
Sorry, so could you bundle up the .exe to the zip file, run from there, then a corresponding savefile is created in a directory based on the zip file name?

**Comment by HiPhish** [comment]
I'm still not sure what you are trying to do.

All the directories passed to PhysicsFS are passed at runtime, so you could use whatever you want to generate the strings. For example, if the version number of Barony was `#define`d you could do something like this:
```c
#define BARONY_VERSION "2.0.5"

/* using the fact that adjacent string literals are concatenated in C */
PHYSFS_setWriteDir("~/.config/Barony/" BARONY_VERSION);
/* This sets the write directory to ""~/.config/Barony/2.0.5" */
```
I don't know how you would run an EXE from inside a ZIP file, but if you have a way of getting the name of the ZIP file at runtime then you could build the string out of that. Once the write directory is set up the rest happens automatically. PhysicsFS is not concerned with *how* you determine the directory name, that part is up to you.

**Comment by WALLOFJUSTICE** [comment]
I'm probably thinking a bit far ahead at this point about how the final bundle of files is shared/looks like...  We can definitely make good use of this system. You're welcome to start setting up the framework when you find the time, happy to review it.

**Comment by HiPhish** [comment]
>I'm probably thinking a bit far ahead at this point about how the final bundle of files is shared/looks like

I was thinking about a setup like this: All the game's official assets are shipped with the binary just as they are now. Maybe they could get ZIPed to make moving assets easier, but that's not strictly necessary. This would be the lowest-priority search path. Then comes a series of "mod directories", first one that is accessible to all users and then one that is local to the current users. On macOS the order would look like this, from lowest to highest priority:
```
. (current directory, i.e. the way it is now, this is what the game ships with)
/Library/Application Support/Barony/Mods/*.zip (global mods)
~/Library/Application Support/Barony/Mods/*.zip (user-specific mods)
```
The `Mods` directories would be searched for archives and the contents of each archive would be added to the VFS. This has the disadvantage that to enable or disable mods one would have to move them in and out of that directory, but it would work. A mod launcher which enables or disables mods like Dungeons of Dredmor has it could be added later.

Save files would be written to the write directory, which could be `~/Library/Application Support/Barony/User/` on macOS.

For starters I would not add any additional search paths, just wrap all the raw file access with the VFS. Once that is done we can think about all the other directories.

**Comment by lheckemann** [comment]
FWIW Barony can look for game data in a different directory using the -datadir option now, since #71 has been merged. I intend to add the same thing for user data. These changes should be useful for reimplementing this using physfs or similar, as at that point only `src/files.cpp` should need modification.

**Comment by HiPhish** [comment]
> FWIW Barony can look for game data in a different directory using the -datadir option now, since #71 has been merged. I intend to add the same thing for user data.

That's cool, if those options are specified we can use the values when initialising PhysicsFS.

>as at that point only `src/files.cpp` should need modification.

I have noticed that there are still file accesses outside of `src/files.cpp`, for example in `src/sound.cpp` the function `openal_oggopen` is using a file name to get a `FILE` handle.
```c++
static int openal_oggopen(OPENAL_SOUND *self, const char* oggfile) {
	FILE *f = fopen(oggfile, "rb");
	//...
}
```
For full PhysicsFS support we need to find every instance where a file is accessed. For raw file access I can just search for `fopen` in the source, but libraries which use file names to find the file on their own need to be made to work with raw memory instead. I already know that SDL_image and SDL_mixer can do that, but I'll have to look into FMOD.

**Comment by lheckemann** [comment]
You're probably looking on the master branch, where @addictgamer reverted the changes introduced by #64 because he couldn't merge them into barony-next, which is where all the new development is taking place (and which is where #71 was merged to). I did remove all occurrences of `fopen` for accessing game data (as opposed to user files) outside `files.cpp`. Functions that handle opening files themselves (e.g. `TTF_OpenFont`) are always preceded by `completePath` in the current code.

It's fairly easy to find the places where a file is accessed by using strace to find the paths that are accessed. If that doesn't help find the code where it's happening, using gdb to set tracepoints on the `open`, `access` and `stat` system calls will.

I don't know what the opinion of the people who have the say in this matter is, but I think it would be good to remove FMOD completely given that OpenAL support works and FMOD causes some other problems (see #19).

**Comment by HiPhish** [comment]
>You're probably looking on the master branch

Yeah, thanks for that info, I'll switch over to `barony-next` then.

>Functions that handle opening files themselves (e.g. `TTF_OpenFont`) are always preceded by `completePath` in the current code.

That's not enough though, in a VFS a "file" might not be an actual file on the file system. It could also be a file inside another file, like inside a ZIP file. For example, Doom's assets are all contained inside a WAD file, so if the engine wants an image it searches the file system for the WAD file, and then inside the WAD file it takes out a slice of it which is the data of the sprite.

The proper way of doing file access with a VFS is to load the raw data of the "file" and then pass that data to the library. In the case of SDL_Image I would do something like this:
```c++
unsigned char *raw_data;
PHYSFS_File *sprite_file = PHYSFS_openRead("foo/bar.png");
PHYSFS_read(sprite, raw_data, 1, PHYSFS_fileLength(sprite_file));
SDL_RWops *data = SDL_RWFromConstMem(raw_data, PHYSFS_fileLength(sprite_file));
PHYSFS_close(sprite_file);
SDL_Surface *sprite = IMG_Load_RW(data, 1);
```
This is the sort of ritual one has to go through with PhysicsFS.

**Comment by lheckemann** [comment]
> That's not enough though, in a VFS a "file" might not be an actual file on the file system.

I realise that, I just mentioned it as it can help find cases where changes will be necessary, however… It turns out that the font loading was the only such case, and I've removed any direct opening of game data files, instead always going through files.cpp: 3524db3891b2d3f35c589cd0cb017c87018ba150

**Comment by HiPhish** [comment]
Cool, no need to hurry though, I won't be able to get into implementing a VFS for a while yet. I'm only dipping my feet in the Barony source code for now. I find it easier to write code than to read other people's code ^^

**Comment by lheckemann** [comment]
Done in fe868eab0e1e97892f6fa49e18d79175a8590eb2. Now I just need to get the user data abstracted, then VFS support should only require modifying files.cpp.

**Comment by lheckemann** [comment]
I've opened #85 which includes all the relevant changes.

**Comment by lheckemann** [comment]
So I never got around to finishing this all the way, which is largely due to fmod being obnoxious. I could probably get it working for only OpenAL.

**Comment by crkellen** [comment]
I currently have a working version of OpenAL for Windows, if you required testing on Windows as well.
null

**Comment by SheridanR** [comment]
Only OpenAL is sufficient -- FMOD support is officially deprecated at this point.

**Comment by lheckemann** [comment]
Awesome, thanks for the input!

**Comment by lheckemann** [comment]
Also — I'm trying to port the user data dir changes to barony-next, but can't test them because the 2.0.5 data isn't compatible with the barony-next branch anymore, even with the [dummy data provided by Addict](https://github.com/TurningWheel/Barony/pull/64#issuecomment-302276146). Any way I can get data that will work with barony-next?

**Comment by crkellen** [comment]
I believe you need the newest version of the dummy assets. Let me see if I can grab the link.
null

**Comment by crkellen** [comment]
Yes, addictgamer has updated the dummy assets, you can grab the newest version here: https://github.com/TurningWheel/Barony/wiki/Barony-Next----Dummy-Assets

That should solve your issue, as I had the same problems, which lead to updating them.

**Comment by lheckemann** [comment]
That fixed it, thanks!

**Comment by addictgamer** [comment]
> Only OpenAL is sufficient -- FMOD support is officially deprecated at this point.

Well, for the most part. I'm not officially supporting OpenAL at the moment, I think there were still some lingering issues floating around, so the releases are still built with FMOD. However, I do plan on phasing out FMOD and begin distributing with OpenAL officially in the coming releases. It just needs more testing first.

**Comment by lheckemann** [comment]
What's in the way for removing FMOD completely? I'd love for this to happen.

**Comment by addictgamer** [comment]
> What's in the way for removing FMOD completely? I'd love for this to happen.

Let's go with "Nothing."

**Comment by WALLOFJUSTICE** [comment]
Since we're now fully released on 'master' branch, where are we at on this datadir /VFS integration?

I'm thinking to be able to read the contents of /mods/test.zip, and have gamesaves/scores etc be created in /mods/test/ to read from. At the moment just to support custom maps and models, nothing like sounds that may need extra support to hotswap

**Comment by WALLOFJUSTICE** [comment]
Implemented PHYSFS virtual file system since.. 3.1.5 I believe
